common/utils/circular_buffer_utility.h

```diff
@@ -0,0 +1,97 @@
+#ifndef COMMON_UTILS_CIRCULAR_BUFFER_UTILITY_H_
+#define COMMON_UTILS_CIRCULAR_BUFFER_UTILITY_H_
+
+#include <cpp_circular_buffer.h>
+
+#include <cstddef>
+
+using std::size_t;
+
+template <typename Type, size_t Capacity>
+class CircularBufferUtility {
+  CircularBuffer<Type, Capacity>* cb;
+
+  /**
+   * Insert a new element into the buffer and remove the oldest element if the
+   * buffer is full.
+   *
+   * @param item: The element to insert into the buffer.
+   * @return bool: Status, true if dequeue and enqueue succeeds, false if either
+   * fails.
+   */
+  * / bool addElement(Type item) {
+    if (this->cb->full()) {
+      if (!this->cb->dequeue(1)) {
+        return false;
+      }
+    }
+    return this->cb->enqueue(item);
+  }
+
+  /**
+   * Get the average of all elements in the buffer.
+   *
+   * @return Type: The average of all elements in the buffer.
+   */
+  Type average() {
+    Type sum = this->sum();
+    return sum / this->cb->count();
+  }
+
+  /**
+   * Get the median of all elements in the buffer.
+   *
+   * @return Type: The median of all elements in the buffer.
+   */
+  Type median() {
+    size_t size = this->cb->count();
+    Type linearData[Capacity];
+
+    for (size_t i = 0; i < size; i++) {
+      linearData[i] = getNthValue(i);
+    }
+
+    std::sort(linearData, linearData + size);
+    return linearData[size / 2];
+  }
+
+  /**
+   * Get the Nth Value object
+   *
+   * @param destination
+   * @param n
+   * @return bool: Status, true if getting the n'th value succeeds, false
+   * otherwise
+   */
+  bool getNthValue(Type* destination, int n) {
+    if (n >= this->cb->count() || n < 0) {
+      return false;
+    }
+
+    if (this->head + n <= endPtr) {
```

Comment:
Because this class isn't extending the circular buffer class and is just using composition, it doesn't have the `head` and `tail` and `backingarray` of the circular buffer. They're also private inside the circular buffer, so they couldn't be accessed here through it, either (i.e. `this->cb->head` will not work because the head is private).

I think that means that this function probably needs to go into the actual circular buffer class and then this class can expose the same function that delegates to the circular buffer.